src/config.rs

```diff
@@ -27,6 +30,49 @@ pub enum ConfigError {
     AmbiguousSource(PathBuf, PathBuf),
 }
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ConfigSource {
+    Default,
+    Env,
+    // TODO: Track explicit file paths when user config is a dir.
+    User,
+    Repo,
+    CommandArg,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum AnnotatedValue {
+    Value {
+        value: config::Value,
+        source: ConfigSource,
+        is_overridden: bool,
+    },
+    Table(HashMap<String, AnnotatedValue>),
+}
+
+impl AnnotatedValue {
+    pub fn iter_recursive<'a>(
+        &'a self,
+        prefix: Option<String>,
+    ) -> Box<dyn Iterator<Item = (Option<String>, &AnnotatedValue)> + 'a> {
+        match *self {
+            AnnotatedValue::Value { .. } => Box::new([(prefix, self)].into_iter()),
+            AnnotatedValue::Table(ref table) => Box::new(
+                // TODO: Remove sorting when config crate maintains deterministic ordering.
+                table
+                    .iter()
+                    .sorted_by_key(|(k, _)| *k)
+                    .flat_map(move |(k, v)| {
+                        let key = prefix
+                            .to_owned()
+                            .map_or(k.to_owned(), |n| format!("{n}.{k}"));
+                        v.iter_recursive(Some(key))
+                    }),
+            ),
+        }
```

Comment:
Not sure I follow this suggestion. Is it still relevant along with other suggested changes to narrow config tables?

Reply:
I meant something like this might be easier to process than recursive structure:
```
[
    (["git", "pull"], Value::String("origin"), Source::Default),
    (["git", "pull"], Value::String("upstream"), Source::User),
    (["ui", "pager"], Value::Array(["less"]), Source::User),
]
```

Default values can be excluded by `.iter().filter(|(.., source)| ...`, and entries that share the same key (or maybe key prefix?) with the latter should be overridden.

I didn't consider that thoroughly, so I might be wrong.

Reply:
SG, went with an approach along those lines and flattened out the repeated recursion.